snuba/query/matchers.py

```diff
@@ -323,4 +323,36 @@ def match(self, node: AnyType) -> Optional[MatchResult]:
         return result
 
 
+@dataclass(frozen=True)
+class TransformedColumn(Pattern[FunctionCallExpr]):
+    """
+    Matches a Column that has been wrapped in zero or more Functions. It recursively
+    checks the parameters of the FunctionCall to see if at least one of them is a Column
+    matching the provided column name.
+
+    Note that this makes no guarantees about the functions being called, it naively
+    goes through all the parameters and returns the first column match it finds.
+    """
+
+    column: Optional[Column] = None
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        if not isinstance(node, (ColumnExpr, FunctionCallExpr)):
+            return None
+
+        if isinstance(node, ColumnExpr) and self.column is not None:
+            return self.column.match(node)
+
+        if isinstance(node, FunctionCallExpr):
```

Comment:
I think we should have a list of allowed functions to be ignored.
`count(toStartOfDay(timestamp)) < 4` (Or any boolean condition).
Should never be confused with a timestamp condition, since, while the `timestamp` column is wrapped in a function that still returns a timestamp. The result is then wrapped into a function that does not have this property.

We could also provide default lists of allowed functions for common use like `TIMESTAMP_FUNCITONS`, `ARITHMETIC_FUNCTIONS`.

Reply:
I will definitely add something to ignore boolean and conditional functions, since this shouldn't match those. But I'm less sure about providing a whitelist of functions. Clickhouse has more than 60 date functions alone, and that's not counting arithmetic functions, generic functions like ifNull etc. Your proposal also assumes that the match cares if the column type is preserved. While that is sometimes true, it's not always true.

I think it's better to not provide any typing guarantees in this function and document it (which I did). Trying to handle every edge case to ensure that the type is matched would be too much work for minimal gain.

Reply:
I don't think I can agree with this approach. 
There are definitively a lot more combinations of functions that would change the type of the column we look for or change the semantics (see the example `count(toStartOfDay(timestamp))`  than functions that are supposed to return the same type (thus we can list). Since we do not have control around how the user composes the query, there would be more chances to match a function we should not match by mistake than ignore an expression we should not ignore.

So we would have false positives and false negatives, but it seems that without listing the allowed functions we would optimize to avoid false negatives making a larger mistake than if we optimized to avoid false positives.

Reply:
OK well I think this new matcher needs to be thought out more, so I'm going to remove it and open it again in a separate PR. I don't want this to block the removal of process_condition.